File: src/main/java/com/donrobo/fpbg/planner/ProductionLine.java
package com.donrobo.fpbg.planner;

import com.donrobo.fpbg.data.Item;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Data
@NoArgsConstructor
public class ProductionLine {

    private final List<String> inputMaterials = new ArrayList<>();
    private final List<ProductionStep> productionSteps = new ArrayList<>();

    public void addProductionStep(ProductionStep productionStep) {
        if (isUsingOnlyInputMaterials(productionStep)) {
            productionSteps.add(0, productionStep);
        } else {
            FractionalItemStack resultsPerSecond = productionStep.getResultPerSecond();
            String result = resultsPerSecond.getItem().getName();
            int index = 0;
            while (index < productionSteps.size() && !isRequiredBy(result, productionSteps.get(index))) {
                index++;
            }
            productionSteps.add(index, productionStep);
        }
    }

    private boolean isRequiredBy(String item, ProductionStep productionStep) {
        return productionStep.getIngredientsPerSecond().stream().map(i -> i.getItem().getName()).anyMatch(ingredient -> ingredient.equals(item));
    }

    private boolean isUsingOnlyInputMaterials(ProductionStep productionStep) {
        return productionStep.getIngredientsPerSecond().stream().map(i -> i.getItem().getName()).allMatch(inputMaterials::contains);
    }


    public void addInputMaterials(List<Item> inputMaterials) {
        inputMaterials.stream().map(Item::getName).forEach(this.inputMaterials::add);
    }

    public void clearUnusedInputMaterials() {
        List<String> usedInputMaterials = inputMaterials.stream().filter(mat ->
                productionSteps.stream().anyMatch(ps -> isRequiredBy(mat, ps))
        ).collect(Collectors.toList());

        inputMaterials.clear();
        inputMaterials.addAll(usedInputMaterials);
    }

    public List<String> getAllIngredients() {
        List<String> allIngredients = new ArrayList<>();

        for (ProductionStep ps : productionSteps) {
            ps.getIngredientsPerSecond().stream().map(i -> i.getItem().getName()).forEach(allIngredients::add);
        }

        return allIngredients.stream().distinct().collect(Collectors.toList());
    }

    public List<ProductionStep> getProductionStepsThatRequire(String input) {
        return productionSteps.stream().filter(ps -> ps.requires(input)).collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/donrobo/fpbg/planner/ProductionLine.kt
package com.donrobo.fpbg.planner

import com.donrobo.fpbg.data.Item
import java.util.*

class ProductionLine {

    private val mutableInputMaterials = ArrayList<String>()
    private val mutableProductionSteps = ArrayList<ProductionStep>()

    fun addProductionStep(productionStep: ProductionStep) {
        if (isUsingOnlyInputMaterials(productionStep)) {
            mutableProductionSteps.add(0, productionStep)
        } else {
            val (_, item) = productionStep.resultPerSecond
            val result = item.name
            var index = 0
            while (index < mutableProductionSteps.size && !isRequiredBy(result, mutableProductionSteps[index])) {
                index++
            }
            mutableProductionSteps.add(index, productionStep)
        }
    }

    fun isRequiredBy(item: String, productionStep: ProductionStep) = productionStep.ingredientsPerSecond.map { it.item.name }.contains(item)

    fun isUsingOnlyInputMaterials(productionStep: ProductionStep) = productionStep.ingredientsPerSecond.map { it.item.name }.all { inputMaterials.contains(it) }

    fun addInputMaterials(inputMaterials: List<Item>) = this.mutableInputMaterials.addAll(inputMaterials.map { it.name })

    fun clearUnusedInputMaterials() {
        val usedInputMaterials = inputMaterials.filter { mat -> productionSteps.any { ps -> isRequiredBy(mat, ps) } }

        mutableInputMaterials.clear()
        mutableInputMaterials.addAll(usedInputMaterials)
    }

    val inputMaterials: List<String> get() = mutableInputMaterials

    val productionSteps: List<ProductionStep> get() = mutableProductionSteps

    val allIngredients: List<String> get() = productionSteps.flatMap { it.ingredientsPerSecond.map { it.item.name } }.distinct()

    fun getProductionStepsThatRequire(input: String): List<ProductionStep> = productionSteps.filter { it.requires(input) }

}
